Temporal Nexus: Deploy to Production - Temporal Cloud & Nexus (#3078)
temporalio/documentation: docs/production-deployment/cloud/nexus/getting-started.mdx

Steps to decompose a large monolithic Namespace:

1. Hide service implementations behind a Nexus `Endpoint`  
   1. Add Nexus `Services` to the same `Workers` as the Temporal primitives being abstracted  
   2. Add Nexus `Endpoints` to the Nexus API Registry, with monolithic Namespace as the target  
2. Service consumers use the Nexus Endpoint instead of the underlying implementation  
   1. Can be done incrementally, until there are no direct caller dependencies on the underlying service implementations (the underlying Temporal primitives)  
3. Move service implementations to a different Namespace  
   1. Create a new `Namespace`  
   2. Add a `Worker` deployment with the Nexus `Service`  
   3. Update the Nexus `Endpoint` target to the new Namespace  
   4. Configure the Endpoint allowlist to allow calls from the original monolithic Namespace  
   5. New Nexus `Operations` will be routed from callers in the monolithic Namespace to the new Namespace  
4. Quiesce Nexus Operations on the old target Namespace  
   1. Leave the old `Worker` deployment running until all existing Nexus Operations in the old Namespace have completed (and their underlying `Workflows`, if any).  

>    2. Old Workers in the monolithic Namespace can have the service implementation removed, since it is now being served from the new `Endpoint` target Namespace.

Author: Can "previous" workers be used instead of "old workers"? @prasek

---
id: getting-started
slug: /cloud/nexus/getting-started
title: Getting Started with Temporal Nexus
description: Get started with Temporal Nexus, including setting up Nexus Endpoints and integrating Nexus into your Temporal Workflows.
sidebar_label: Getting Started
tags:
  - temporal cloud
  - nexus setup
keywords:
  - Temporal Nexus onboarding
  - setting up Nexus Endpoints
  - temporal cloud
  - nexus setup

---

Temporal Nexus Public Preview is available within your Temporal Cloud account, with support in the [Temporal Go SDK](/develop/go).

Calls across existing Namespaces can be enabled by creating a Nexus Endpoint in the Nexus API Registry, creating a Nexus Service in a Worker in the handler Namespace, and then using the Nexus Service from a caller Workflow in a different Namespace.

> Monolithic Namespaces can be decomposed into multiple Namespaces, by hiding service implementations behind a Nexus Endpoint in the monolithic Namespace, pointing all consumers at the new Nexus Endpoint, and then changing the Endpoint’s target Namespace to a different Namespace. Multiple Nexus Endpoints can target a single monolithic Namespace.

Reviewer: This is really confusing and I don't understand the why behind this. Is the goal that there's one entry point that can be broken down to simple components and treated as a single endpoint, so you don't  have to use a monolithic implementation? I _think_ this is sort of what you're aiming for.

```suggestion
Monolithic Namespaces can be decomposed into multiple Namespaces by hiding service implementations behind  Nexus Endpoints.
This points consumers to a single Nexus Endpoint, redirecting work to a specialized Namespace.
This allows each subordinate Namespace to specialize its functionality without creating monoliths.
Nexus Endpoint redirects are not exclusive.
More than one Nexus Endpoint can redirect to the same subordinate Namespace, allowing Namespace functionality to be shared between Nexus solutions.
```

## Cross-Namespace

> Calls across existing Namespaces can be enabled by creating a Nexus Endpoint in the Nexus API Registry, creating a Nexus Service in a Worker in the handler Namespace, and then using the Nexus Service from a caller Workflow in a different Namespace.

Reviewer: ```suggestion
Creating a Nexus Endpoint in the Nexus API registry enables calls across existing Namespaces.
With Nexus Services, you can create a Worker in a handler Namespace and a calling Workflow in a separate Namespace.
```

Confused. Again, I think this is what this is supposed to mean.

> Steps to enable calls across existing Namespaces:

Reviewer: ```suggestion
Follow these steps to enable calls across Namespaces:
```

> 1. Add Nexus Services to the same Workers as the Temporal primitives being abstracted. 

Reviewer: ```suggestion
1. A Temporal primitive is a Workflow, Activity, or Worker.
   Add Nexus Services to each Worker that hosts primitives you need to make visible.
```

Author: I'm not 100% on if we should take this in so I'm resolving it.

2. Add a Nexus Endpoint that:  

>    1. Targets the handler Namespace. 

Reviewer: ```suggestion
   1. Points to ("targets") the Namespace that hosts the Workflow Executions and Activity Executions that handle the required work.
```

>    2. Allows the caller Namespace.  

Reviewer: Again, guessing on this

```suggestion
   2. Lists the Namespaces ("callers") that may place requests to the handler Namespaces.
```

Author: not sure if this is correct so resolving it.

> 3. Make Nexus calls from a caller Workflow in a different Namespace.  

Reviewer: ```suggestion
3. To make Nexus calls from a caller Workflow that lives in a different Namespace:
```

Author: Not 100% on this. Resolving.

>    1. Use `workflow.NewNexusClient(endpointName, serviceName)`.  

Reviewer: ```suggestion
   1. Create a Nexus-enabled Client to communicate across Namespaces with `workflow.NewNexusClient(endpointName, serviceName)`.
```

   2. Execute a Nexus Operation: `nexusClient.ExecutionOperation(...)`

## Decompose a Monolithic Namespace

> Multiple Nexus Endpoints can target a single monolithic Namespace, and then each Endpoint can be updated, one at a time, to target separate Namespaces, for an incremental migration. 

Reviewer: ```suggestion
Nexus enables you to simplify monolithic Namespaces and decompose them into logically focused areas of responsibility.
Decomposition can be a multi-stage process. 
To start, multiple Nexus Endpoints can target a single monolithic Namespace.
Then you can update each Endpoint to target separated Namespaces for your incremental migration.
```

> Once Nexus Endpoints are in place, targeting a new Namespace can be done with config changes and zero downtime, as new Nexus requests will be routed to the new target Namespace, and existing Nexus requests will be completed in the previous Namespace.

Reviewer: ```suggestion
Once Nexus Endpoints are in place, targeting a new Namespace can be applied using config changes.
This involves zero downtime.
New Nexus requests will be routed to the new target Namespace, and existing Nexus requests will be completed in the previous (current) Namespace.
```

**Steps to decompose a large monolithic Namespace:**

> 1. Hide service implementations behind a Nexus Endpoint. 

Reviewer: ```suggestion
1. Create a single Nexus Endpoint.
   This hides service implementations, and allows you to incrementally break down your monolith.
```

>    1. Add Nexus Services to the same Workers as the Temporal primitives being abstracted. 

Reviewer: ```suggestion
   1. Add Nexus Services to the Workers that host the Workflows and Activities you'll need to access. 
```

>    2. Add Nexus Endpoints to the Nexus API Registry, with monolithic Namespace as the target. 

Reviewer: ```suggestion
   2. Add each Nexus Endpoint to the Nexus API Registry using the monolithic Namespace as the initial target. 
```

> 2. Service consumers use the Nexus Endpoint instead of the underlying implementation.  

Reviewer: ```suggestion
2. Migrate your Service consumers to use the Nexus Endpoint instead of the underlying implementation.  
```

>    1. This can be done incrementally, until there are no direct caller dependencies on the underlying service implementations (the underlying Temporal primitives). 

Reviewer: ```suggestion
   This can be done incrementally, until there are no direct caller dependencies on the underlying service implementations, that is the underlying Temporal Workflow and Activity primitives.
```

> 3. Move service implementations to a different Namespace.

Reviewer: ```suggestion
3. Decouple your Service implementations to new Namespaces, where they are grouped in a more functional manner instead of a monolith.
```

Author: Going to resolve these. As discussed, a lot of this will end up needed rewriting in another iteration to fit our style (i.e - reading level, clarity, etc).

   1. Create a new Namespace.
   2. Add a Worker deployment with the Nexus Service.  
   3. Update the Nexus Endpoint target to the new Namespace. 

>    4. Configure the Endpoint allowlist to allow calls from the original monolithic Namespace.  

Reviewer: ```suggestion
   4. Configure the Endpoint `allowlist` to permit calls from the original monolithic Namespace.  
```

Author: I'm not sure about allowlist. I'll make a note to ask.

>    5. New Nexus Operations will be routed from callers in the monolithic Namespace to the new Namespace.  

Reviewer: ```suggestion
      New Nexus Operations will be routed from callers in the monolithic Namespace to the new Namespace.  
```

> 4. Quiesce Nexus Operations on the previous target Namespace.

Reviewer: ```suggestion
4. Prepare Nexus Operations on the previous target Namespace during the migration.
```

Author: @prasek What's another word that we can use as an alternative to Quiesce that would be accurate to this step?

>    1. Leave the previous Worker deployment running until all existing Nexus Operations in the previous Namespace have completed (and their underlying Workflows, if any).  

Reviewer: ```suggestion
   1. Leave the previous Worker deployment running until all existing Nexus Operations in the previous Namespace have completed.
      This includes their underlying Workflows, if they exist.
```

Author: Took this in.

>    2. Previous Workers in the monolithic Namespace can have the service implementation removed, since it is now being served from the new Endpoint target Namespace.

Reviewer: ```suggestion
   2. Remove Workers from the monolithic Namespace, since those Workflows and Activities are now served from the new Endpoint targeted Namespace.
```

> There are a few caveats:

Reviewer: ```suggestion
Please note the following points:
```

>    1. For long-lived entity Workflows, new Nexus requests will be routed to the new target Namespace and the long-lived entity-Workflow will still be running on the previous Namespace.

Reviewer: ```suggestion
   - For long-lived entity Workflows, new Nexus requests will be routed to the new target Namespace. 
      Long-lived entity-Workflow will still be running on the previous Namespace.
```

>    2. If using Signal-with-Start behind a Nexus Operation, then a new entity-Workflow will be started in the new target Namespace.

Reviewer: ```suggestion
  - When using Signal-with-Start behind a Nexus Operation, a new entity-Workflow will start in the new target Namespace.
```

>    3. A cancel Nexus Operation request from a caller Workflow will be routed through the Nexus Endpoint to the new target Namespace, which may not have the underlying Workflow.

Reviewer: ```suggestion
   - A cancel Nexus Operation request from a caller Workflow will be routed through the Nexus Endpoint to the new target Namespace.
     The new Namespace may not have the underlying Workflow.
```

> In these scenarios, Temporal can provide guidance on different migration strategies.

Reviewer: How should someone seek this guidance? Support? Community?

---
id: getting-started
slug: /cloud/nexus/getting-started
title: Getting Started with Temporal Nexus
description: Learn how to get started with Temporal Nexus, including setting up Nexus Endpoints and integrating Nexus into your Temporal workflows.
sidebar_label: Getting Started
tags:
  - temporal cloud
  - workflow onboarding
  - nexus setup
  - API registry
  - cloud workflows
keywords:
  - Temporal Nexus onboarding
  - setting up Nexus Endpoints
  - workflow onboarding
  - cross-namespace integration
  - Nexus API registry
  - starting with Temporal Nexus
---

#### Getting Started

##### Overview

Temporal Nexus Public Preview is available within your Temporal Cloud account, with support in the Temporal Go SDK ([developer guide]) and Temporal Java SDK (developer guide).

> Calls across existing namespaces can be enabled by creating a Nexus Endpoint in the Nexus API Registry, creating a Nexus Service in a Worker in the handler namespace, and then using the Nexus Service from a caller Workflow in a different namespace.

Reviewer: uppercase Namespace

---
id: getting-started
slug: /cloud/nexus/getting-started
title: Getting Started with Temporal Nexus
description: Get started with Temporal Nexus, including setting up Nexus Endpoints and integrating Nexus into your Temporal Workflows.
sidebar_label: Getting Started
tags:
  - temporal cloud
  - nexus setup
keywords:
  - Temporal Nexus onboarding
  - setting up Nexus Endpoints
  - temporal cloud
  - nexus setup

---

Temporal Nexus Public Preview is available within your Temporal Cloud account, with support in the [Temporal Go SDK](/develop/go).

Calls across existing Namespaces can be enabled by creating a Nexus Endpoint in the Nexus API Registry, creating a Nexus Service in a Worker in the handler Namespace, and then using the Nexus Service from a caller Workflow in a different Namespace.

Monolithic Namespaces can be decomposed into multiple Namespaces, by hiding service implementations behind a Nexus Endpoint in the monolithic Namespace, pointing all consumers at the new Nexus Endpoint, and then changing the Endpoint’s target Namespace to a different Namespace.

> Multiple Nexus Endpoints can target a single monolithic Namespace.

Reviewer 2: [Nit] "can target a single Namespace" --> I think monolithic is not needed for this line
